File: frontend/src/api/usage-analytics.ts
import { get } from '@/utils/request'

export interface UsageQuery {
  from?: string
  to?: string
  tenant_id?: number
  interval?: 'hour' | 'day' | 'week' | 'month'
  page?: number
  page_size?: number
  sort?: string
  operation?: string
  usage_class?: 'foreground' | 'background'
  channel?: string
  model_type?: string
  direction?: 'inbound' | 'outbound'
  group_by?: 'caller' | 'knowledge_base' | 'tool' | 'client'
  status?: UsageGovernanceStatus
  include_inactive?: boolean
  mcp_adoption?: 'adopted' | 'not_adopted'
  cross_tenant?: 'with' | 'without'
  owner_tenant_id?: number
  knowledge_base_id?: string
  metric?: 'tokens' | 'accesses'
  dimension?: 'total' | 'source' | 'scope'
}

export type UsageGovernanceStatus = 'active' | 'low_activity' | 'inactive' | 'never_used' | 'insufficient_data'

export interface UsageAttention {
  code: string
  entity_type?: string
  entity_id?: string
  entity_name?: string
  value?: number
}

export interface UsageOverview {
  total_tokens: number
  foreground_tokens: number
  background_tokens: number
  input_tokens: number
  output_tokens: number
  cache_read_tokens: number
  cache_write_tokens: number
  assistant_turns: number
  mcp_calls: number
  mcp_success_rate: number
  mcp_average_latency_ms: number
  unattributed_mcp_calls: number
  active_tenants: number
  active_principals: number
  active_knowledge_bases: number
  cross_tenant_usage: number
  unattributed_mcp_ratio: number
  inactive_knowledge_bases: number
  mcp_active_tenants: number
  mcp_platform_penetration: number
  mcp_adoption_among_active_spaces: number
  token_growth_percent: number
  top_knowledge_bases: KnowledgeBaseUsageRow[]
  attention_needed: UsageAttention[]
  collecting_since?: string
}

export interface UsagePage<T> {
  data: T[]
  page: number
  page_size: number
  total: number
  collecting_since?: string
}

export interface TenantUsageRow {
  tenant_id: number
  tenant_name: string
  active_principals: number
  assistant_turns: number
  agent_turns: number
  input_tokens: number
  output_tokens: number
  total_tokens: number
  mcp_calls: number
  mcp_success_rate: number
  mcp_avg_latency_ms: number
  last_active?: string
  usage_status: UsageGovernanceStatus
  kb_used_count: number
  kb_owned_count: number
  external_kb_used_count: number
  cross_tenant_accesses: number
  owned_kb_cross_tenant_accesses: number
  mcp_adopted: boolean
}

export interface UsageTimeSeriesPoint {
  bucket: string
  input_tokens: number
  output_tokens: number
  total_tokens: number
  assistant_turns: number
  mcp_calls: number
  model_accesses: number
  internal_accesses: number
  external_accesses: number
}

export interface ModelUsageRow {
  model_id: string
  model_type: string
  assistant_turns: number
  input_tokens: number
  output_tokens: number
  total_tokens: number
  cache_read_tokens: number
  cache_write_tokens: number
  last_active?: string
}

export interface OperationUsageRow {
  operation: string
  usage_class: 'foreground' | 'background'
  invocations: number
  total_tokens: number
  percentage: number
}

export interface MCPUsageRow {
  tool_name: string
  calls: number
  successful_calls: number
  success_rate: number
  average_latency_ms: number
  unattributed_calls: number
  last_active?: string
  client_name?: string
  client_version?: string
  knowledge_base_id?: string
  knowledge_base_name?: string
}

export interface KnowledgeBaseUsageRow {
  knowledge_base_id: string
  knowledge_base_name: string
  owner_tenant_id: number
  owner_tenant_name: string
  caller_tenant_id?: number
  caller_tenant_name?: string
  assistant_turns: number
  mcp_calls: number
  last_active?: string
  usage_status: UsageGovernanceStatus
  total_accesses: number
  model_accesses: number
  unique_tenants: number
  unique_principals: number
  external_tenants: number
  cross_tenant_accesses: number
  cross_tenant_share: number
  recent_growth: number
}

function params(query: UsageQuery): string {
  const value = new URLSearchParams()
  Object.entries(query).forEach(([key, item]) => {
    if (item !== undefined && item !== '') value.set(key, String(item))
  })
  const encoded = value.toString()
  return encoded ? `?${encoded}` : ''
}

const base = '/api/v1/system/admin/usage'

export const getUsageOverview = (query: UsageQuery) => get(`${base}/overview${params(query)}`) as Promise<UsageOverview>
export const getUsageTenants = (query: UsageQuery) =>
  get(`${base}/tenants${params(query)}`) as Promise<UsagePage<TenantUsageRow>>
export const getUsageTimeSeries = (query: UsageQuery) =>
  get(`${base}/timeseries${params(query)}`) as Promise<UsagePage<UsageTimeSeriesPoint>>
export const getUsageModels = (query: UsageQuery) =>
  get(`${base}/models${params(query)}`) as Promise<UsagePage<ModelUsageRow>>
export const getUsageOperations = (query: UsageQuery) =>
  get(`${base}/operations${params(query)}`) as Promise<UsagePage<OperationUsageRow>>
export const getMCPUsage = (query: UsageQuery) => get(`${base}/mcp${params(query)}`) as Promise<UsagePage<MCPUsageRow>>
export const getKnowledgeBaseUsage = (query: UsageQuery) =>
  get(`${base}/knowledge-bases${params(query)}`) as Promise<UsagePage<KnowledgeBaseUsageRow>>

File: internal/types/usage_analytics.go
package types

import (
	"crypto/sha256"
	"encoding/hex"
	"strings"
	"time"
)

const (
	UsageEventKindModel   = "model"
	UsageEventKindMCP     = "mcp"
	UsagePrincipalUnknown = "system/unknown"
)

const (
	UsageClassForeground = "foreground"
	UsageClassBackground = "background"

	ModelUsageOperationKnowledgeQA           = "knowledge_qa_turn"
	ModelUsageOperationAgent                 = "agent_turn"
	ModelUsageOperationAgentCompaction       = "agent_compaction"
	ModelUsageOperationQueryRewrite          = "query_rewrite"
	ModelUsageOperationEntityExtraction      = "entity_extraction"
	ModelUsageOperationDataAnalysisPlanning  = "data_analysis_planning"
	ModelUsageOperationDocumentSummary       = "document_summary"
	ModelUsageOperationGeneratedQuestions    = "generated_questions"
	ModelUsageOperationAutoTag               = "auto_tag"
	ModelUsageOperationGraphExtraction       = "graph_extraction"
	ModelUsageOperationSpreadsheetMetadata   = "spreadsheet_metadata"
	ModelUsageOperationSessionTitle          = "session_title"
	ModelUsageOperationMemoryExtraction      = "memory_extraction"
	ModelUsageOperationMemoryConsolidation   = "memory_consolidation"
	ModelUsageOperationMemoryTopicResolution = "memory_topic_resolution"
	ModelUsageOperationWikiIngestion         = "wiki_ingestion"
	ModelUsageOperationWikiGeneration        = "wiki_generation"
	ModelUsageOperationWikiModification      = "wiki_modification"
	ModelUsageOperationMCPInstruction        = "mcp_instruction_generation"
	ModelUsageOperationEmbedding             = "embedding"
	ModelUsageOperationRerank                = "rerank"
	ModelUsageOperationVLM                   = "vlm"
	ModelUsageOperationASR                   = "asr"
)

var foregroundModelUsageOperations = map[string]struct{}{
	ModelUsageOperationKnowledgeQA: {},
	ModelUsageOperationAgent:       {},
}

var supportedModelUsageOperations = map[string]struct{}{
	ModelUsageOperationKnowledgeQA: {}, ModelUsageOperationAgent: {},
	ModelUsageOperationAgentCompaction: {}, ModelUsageOperationQueryRewrite: {},
	ModelUsageOperationEntityExtraction: {}, ModelUsageOperationDataAnalysisPlanning: {},
	ModelUsageOperationDocumentSummary: {}, ModelUsageOperationGeneratedQuestions: {},
	ModelUsageOperationAutoTag: {}, ModelUsageOperationGraphExtraction: {},
	ModelUsageOperationSpreadsheetMetadata: {}, ModelUsageOperationSessionTitle: {},
	ModelUsageOperationMemoryExtraction: {}, ModelUsageOperationMemoryConsolidation: {},
	ModelUsageOperationMemoryTopicResolution: {}, ModelUsageOperationWikiIngestion: {},
	ModelUsageOperationWikiGeneration: {}, ModelUsageOperationWikiModification: {},
	ModelUsageOperationMCPInstruction: {}, ModelUsageOperationEmbedding: {},
	ModelUsageOperationRerank: {}, ModelUsageOperationVLM: {}, ModelUsageOperationASR: {},
}

func IsModelUsageOperation(operation string) bool {
	_, ok := supportedModelUsageOperations[operation]
	return ok
}

func ModelUsageClass(operation string) string {
	if _, ok := foregroundModelUsageOperations[operation]; ok {
		return UsageClassForeground
	}
	return UsageClassBackground
}

// UsageEventKey returns a deterministic, bounded key without retaining prompt
// or response content. Callers provide stable business IDs or safe hashes.
func UsageEventKey(operation string, parts ...string) string {
	operation = strings.TrimSpace(operation)
	h := sha256.New()
	for _, part := range parts {
		h.Write([]byte{0})
		h.Write([]byte(strings.TrimSpace(part)))
	}
	return operation + ":" + hex.EncodeToString(h.Sum(nil))
}

// ModelUsageRecordRequest is the privacy-filtered input to the centralized
// model ledger recorder. It deliberately cannot carry prompts or responses.
type ModelUsageRecordRequest struct {
	EventKey         string
	TenantID         uint64
	Channel          string
	Operation        string
	ModelID          string
	ModelType        string
	Usage            TokenUsage
	SessionID        string
	MessageID        string
	RequestID        string
	TraceID          string
	Status           string
	UsageSource      string
	KnowledgeBaseIDs []string
	KnowledgeIDs     []string
}

type MCPUsageRecordRequest struct {
	EventKey         string
	MCPServiceID     string
	ToolName         string
	Transport        string
	Success          bool
	ErrorCode        string
	LatencyMs        int64
	RequestID        string
	TraceID          string
	KnowledgeBaseIDs []string
	KnowledgeIDs     []string
}

// ModelUsageEvent is the append-only analytics representation of one model
// invocation. It intentionally contains no prompt or response content.
type ModelUsageEvent struct {
	ID               uint64    `json:"id" gorm:"primaryKey;autoIncrement"`
	EventKey         string    `json:"event_key" gorm:"type:varchar(255);not null;uniqueIndex"`
	TenantID         uint64    `json:"tenant_id" gorm:"not null;index"`
	PrincipalType    string    `json:"principal_type" gorm:"type:varchar(64);not null"`
	PrincipalID      string    `json:"principal_id,omitempty" gorm:"type:varchar(512)"`
	Channel          string    `json:"channel" gorm:"type:varchar(50);not null"`
	Operation        string    `json:"operation" gorm:"type:varchar(64);not null"`
	ModelID          string    `json:"model_id,omitempty" gorm:"type:varchar(64)"`
	ModelType        string    `json:"model_type,omitempty" gorm:"type:varchar(32)"`
	SessionID        string    `json:"session_id,omitempty" gorm:"type:varchar(36)"`
	MessageID        string    `json:"message_id,omitempty" gorm:"type:varchar(36)"`
	InputTokens      int64     `json:"input_tokens" gorm:"not null"`
	OutputTokens     int64     `json:"output_tokens" gorm:"not null"`
	TotalTokens      int64     `json:"total_tokens" gorm:"not null"`
	CacheReadTokens  int64     `json:"cache_read_tokens" gorm:"not null"`
	CacheWriteTokens int64     `json:"cache_write_tokens" gorm:"not null"`
	ReasoningTokens  int64     `json:"reasoning_tokens" gorm:"not null"`
	UsageSource      string    `json:"usage_source" gorm:"type:varchar(16);not null"`
	Status           string    `json:"status" gorm:"type:varchar(16);not null"`
	RequestID        string    `json:"request_id,omitempty" gorm:"type:varchar(255)"`
	TraceID          string    `json:"trace_id,omitempty" gorm:"type:varchar(255)"`
	OccurredAt       time.Time `json:"occurred_at" gorm:"not null"`
	CreatedAt        time.Time `json:"created_at" gorm:"not null"`
}

func (ModelUsageEvent) TableName() string { return "model_usage_events" }

// MCPUsageEvent represents exactly one inbound MCP tools/call invocation.
type MCPUsageEvent struct {
	ID             uint64    `json:"id" gorm:"primaryKey;autoIncrement"`
	EventKey       string    `json:"event_key" gorm:"type:varchar(255);not null;uniqueIndex"`
	CallerTenantID *uint64   `json:"caller_tenant_id,omitempty" gorm:"index"`
	PrincipalType  string    `json:"principal_type" gorm:"type:varchar(64);not null"`
	PrincipalID    string    `json:"principal_id,omitempty" gorm:"type:varchar(512)"`
	APIKeyID       *uint64   `json:"api_key_id,omitempty"`
	Direction      string    `json:"direction" gorm:"type:varchar(16);not null"`
	MCPServiceID   string    `json:"mcp_service_id,omitempty" gorm:"type:varchar(64)"`
	ToolName       string    `json:"tool_name" gorm:"type:varchar(255);not null"`
	ClientName     string    `json:"client_name,omitempty" gorm:"type:varchar(128)"`
	ClientVersion  string    `json:"client_version,omitempty" gorm:"type:varchar(64)"`
	Transport      string    `json:"transport,omitempty" gorm:"type:varchar(16)"`
	Success        bool      `json:"success" gorm:"not null"`
	ErrorCode      string    `json:"error_code,omitempty" gorm:"type:varchar(64)"`
	LatencyMs      int64     `json:"latency_ms" gorm:"not null"`
	RequestID      string    `json:"request_id,omitempty" gorm:"type:varchar(255)"`
	TraceID        string    `json:"trace_id,omitempty" gorm:"type:varchar(255)"`
	OccurredAt     time.Time `json:"occurred_at" gorm:"not null"`
	CreatedAt      time.Time `json:"created_at" gorm:"not null"`
}

func (MCPUsageEvent) TableName() string { return "mcp_usage_events" }

type UsageResourceLink struct {
	ID               uint64 `json:"id" gorm:"primaryKey;autoIncrement"`
	EventKind        string `json:"event_kind" gorm:"type:varchar(16);not null"`
	EventID          uint64 `json:"event_id" gorm:"not null"`
	ResourceType     string `json:"resource_type" gorm:"type:varchar(32);not null"`
	ResourceID       string `json:"resource_id" gorm:"type:varchar(64);not null"`
	ResourceTenantID uint64 `json:"resource_tenant_id" gorm:"not null"`
}

func (UsageResourceLink) TableName() string { return "usage_resource_links" }

// MCPUsageReport is the deliberately narrow DTO accepted from the Python MCP
// adapter. Identity and tenant fields are absent: the Go authentication
// context is the only authority for them.
type MCPUsageReport struct {
	EventKey         string   `json:"event_key"`
	ToolName         string   `json:"tool_name"`
	ClientName       string   `json:"client_name,omitempty"`
	ClientVersion    string   `json:"client_version,omitempty"`
	Transport        string   `json:"transport,omitempty"`
	SharedGateway    bool     `json:"-"`
	Success          bool     `json:"success"`
	ErrorCode        string   `json:"error_code,omitempty"`
	LatencyMs        int64    `json:"latency_ms"`
	RequestID        string   `json:"request_id,omitempty"`
	KnowledgeBaseIDs []string `json:"knowledge_base_ids,omitempty"`
	KnowledgeIDs     []string `json:"knowledge_ids,omitempty"`
}

type UsageTimeRange struct {
	From       time.Time
	To         time.Time
	TenantID   *uint64
	Interval   string
	Page       int
	PageSize   int
	Sort       string
	Operation  string
	UsageClass string
	Channel    string
	ModelType  string
	Direction  string
}

type UsageOverview struct {
	TotalTokens          int64      `json:"total_tokens"`
	ForegroundTokens     int64      `json:"foreground_tokens"`
	BackgroundTokens     int64      `json:"background_tokens"`
	InputTokens          int64      `json:"input_tokens"`
	OutputTokens         int64      `json:"output_tokens"`
	CacheReadTokens      int64      `json:"cache_read_tokens"`
	CacheWriteTokens     int64      `json:"cache_write_tokens"`
	AssistantTurns       int64      `json:"assistant_turns"`
	MCPCalls             int64      `json:"mcp_calls"`
	MCPSuccessRate       float64    `json:"mcp_success_rate"`
	MCPAverageLatencyMs  float64    `json:"mcp_average_latency_ms"`
	UnattributedMCPCalls int64      `json:"unattributed_mcp_calls"`
	ActiveTenants        int64      `json:"active_tenants"`
	ActivePrincipals     int64      `json:"active_principals"`
	CollectingSince      *time.Time `json:"collecting_since,omitempty"`
}

type TenantUsageRow struct {
	TenantID         uint64     `json:"tenant_id"`
	TenantName       string     `json:"tenant_name"`
	ActivePrincipals int64      `json:"active_principals"`
	AssistantTurns   int64      `json:"assistant_turns"`
	AgentTurns       int64      `json:"agent_turns"`
	InputTokens      int64      `json:"input_tokens"`
	OutputTokens     int64      `json:"output_tokens"`
	TotalTokens      int64      `json:"total_tokens"`
	MCPCalls         int64      `json:"mcp_calls"`
	MCPSuccessRate   float64    `json:"mcp_success_rate"`
	MCPAvgLatencyMs  float64    `json:"mcp_avg_latency_ms"`
	LastActive       *time.Time `json:"last_active,omitempty"`
	TotalCount       int64      `json:"-"`
}

type UsageTimeSeriesPoint struct {
	Bucket         time.Time `json:"bucket"`
	InputTokens    int64     `json:"input_tokens"`
	OutputTokens   int64     `json:"output_tokens"`
	TotalTokens    int64     `json:"total_tokens"`
	AssistantTurns int64     `json:"assistant_turns"`
	MCPCalls       int64     `json:"mcp_calls"`
}

type ModelUsageRow struct {
	ModelID          string     `json:"model_id"`
	ModelType        string     `json:"model_type"`
	AssistantTurns   int64      `json:"assistant_turns"`
	InputTokens      int64      `json:"input_tokens"`
	OutputTokens     int64      `json:"output_tokens"`
	TotalTokens      int64      `json:"total_tokens"`
	CacheReadTokens  int64      `json:"cache_read_tokens"`
	CacheWriteTokens int64      `json:"cache_write_tokens"`
	LastActive       *time.Time `json:"last_active,omitempty"`
	TotalCount       int64      `json:"-"`
}

type OperationUsageRow struct {
	Operation   string  `json:"operation"`
	UsageClass  string  `json:"usage_class"`
	Invocations int64   `json:"invocations"`
	TotalTokens int64   `json:"total_tokens"`
	Percentage  float64 `json:"percentage"`
}

type MCPUsageRow struct {
	ToolName          string     `json:"tool_name"`
	Calls             int64      `json:"calls"`
	SuccessfulCalls   int64      `json:"successful_calls"`
	SuccessRate       float64    `json:"success_rate"`
	AverageLatencyMs  float64    `json:"average_latency_ms"`
	UnattributedCalls int64      `json:"unattributed_calls"`
	LastActive        *time.Time `json:"last_active,omitempty"`
	TotalCount        int64      `json:"-"`
}

type KnowledgeBaseUsageRow struct {
	KnowledgeBaseID   string     `json:"knowledge_base_id"`
	KnowledgeBaseName string     `json:"knowledge_base_name"`
	OwnerTenantID     uint64     `json:"owner_tenant_id"`
	OwnerTenantName   string     `json:"owner_tenant_name"`
	CallerTenantID    *uint64    `json:"caller_tenant_id,omitempty"`
	CallerTenantName  string     `json:"caller_tenant_name,omitempty"`
	AssistantTurns    int64      `json:"assistant_turns"`
	MCPCalls          int64      `json:"mcp_calls"`
	LastActive        *time.Time `json:"last_active,omitempty"`
	TotalCount        int64      `json:"-"`
}

type UsagePage[T any] struct {
	Data            []T        `json:"data"`
	Page            int        `json:"page"`
	PageSize        int        `json:"page_size"`
	Total           int64      `json:"total"`
	CollectingSince *time.Time `json:"collecting_since,omitempty"`
}
